src/typechecker.erl

```diff
@@ -1531,8 +1531,17 @@ add_type_pat_tuple(Pats, {type, _, union, Tys}, TEnv, VEnv) ->
 			    , length(TS) == length(Pats)])),
     lists:foldl(fun ({Pat, Union}, Env) ->
 			add_type_pat(Pat, Union, TEnv, Env)
-		end, VEnv, lists:zip(Pats, Unions)).
-
+                end, VEnv, lists:zip(Pats, Unions));
+add_type_pat_tuple(Pats, Ty, TEnv, VEnv)
+  when element(1, Ty) =:= user_type;
+       element(1, Ty) =:= remote_type ->
+    case normalize(Ty, TEnv) of
```

Comment:
I think normalizing Ty should be done earlier. I think it applies for most clauses of `add_type_pat`, at least for tuple, binary, cons and record. Perhaps it can be done for all these cases of `add_type_pat` using a recursive clause which is placed before the clauses for tuple, cons, bin and record:

```Erlang
add_type_pat(Pat, Ty, TEnv, VEnv)
  when element(1, Ty) =:= user_type;
       element(1, Ty) =:= remote_type ->
    case normalize(Ty, TEnv) of
        Ty ->
            %% no change, unknown case
            throw({unsupported_type, Ty});
        Ty2 ->
            add_type_pat(Pat, Ty2, TEnv, VEnv)
    end;
```